core/src/main/java/com/oath/oak/EntrySet.java

```diff
@@ -473,274 +387,137 @@ boolean casNextEntryIndex(int ei, int nextOld, int nextNew) {
+                ThreadContext.ValueState.MARKED_DELETED_NOT_FINALIZED;
+            return;
         }
-        ValueUtils.ValueResult result = valOffHeapOperator.isValueDeleted(valueSlice, valueSlice.getVersion());
+
+        ValueUtils.ValueResult result = valOffHeapOperator.isValueDeleted(ctx.value);
         if (result == TRUE) {
             // There is a deleted value associated with the given key
-            return new LookUp(
-                null, valueReference, ei, valueSlice.getVersion(), getKeyReference(ei),
-                true);
+            ctx.valueState = ThreadContext.ValueState.MARKED_DELETED_NOT_FINALIZED;
+        } else {
+            // If result == RETRY, we ignore it, since it will be discovered later down the line as well
+            ctx.valueState = ThreadContext.ValueState.VALID;
         }
-        // If result == RETRY, we ignore it, since it will be discovered later down the line as well
-        return new LookUp(valueSlice, valueReference, ei, valueSlice.getVersion(), getKeyReference(ei), false);
     }
 
     /**
      * writeKey writes given key object "key" (to off-heap) as a serialized key, referenced by entry
      * at the entries index "ei"
      **/
-    private void writeKey(K key, int ei) {
+    private void writeKey(ThreadContext ctx, K key, int ei) {
         int keySize = keySerializer.calculateSize(key);
         int intIdx = entryIdx2intIdx(ei);
-        Slice s = memoryManager.allocateSlice(keySize, MemoryManager.Allocate.KEY);
+        memoryManager.allocate(ctx.key, keySize, MemoryManager.Allocate.KEY);
+        ctx.key.setReference(KEY.encode(ctx.key));
         // byteBuffer.slice() is set so it protects us from the overwrites of the serializer
         // TODO: better serializer need to be given OakWBuffer and not ByteBuffer
-        keySerializer.serialize(key, s.getByteBuffer().slice());
-
-        setEntryFieldInt(intIdx, OFFSET.KEY_BLOCK, s.getBlockID());
-        setEntryFieldInt(intIdx, OFFSET.KEY_POSITION, s.getOriginalPosition());
-        setEntryFieldInt(intIdx, OFFSET.KEY_LENGTH, keySize);
+        keySerializer.serialize(key, ctx.key.getAllocByteBuffer().slice());
+        ctx.entryIndex = ei;
+        setEntryFieldLong(intIdx, OFFSET.KEY_REFERENCE, ctx.key.getReference());
     }
 
     /**
-     * readKey reads a key from entry at the given entry index (from off-heap).
-     * Key is returned via reusable thread-local ByteBuffer.
-     * There is no copy just a special ByteBuffer for a single key.
+     * Reads a key from entry at the given entry index (from off-heap).
+     * Returns false if:
+     *   (1) there is no such entry or
+     *   (2) entry has no key set
      * The thread-local ByteBuffer can be reused by different threads, however as long as
      * a thread is invoked the ByteBuffer is related solely to this thread.
      */
-    ByteBuffer readKey(int ei) {
+    boolean readKey(KeyBuffer key, int ei) {
```

Comment:
It bothers me that sometimes entry index (ei) is given via input parameter and sometimes from thread context. It is a good way to the situation when they will be inconsistent an cause a bug one day... There need to be some convention. For example, thread context's ei is set as soon as relevant entry is found or allocated and then only thread context ei is in use. This should eliminate all "ei"'s as inputs. And ThreadCotext transfered everywhere. Or some other rule to define, but to be consistent.

Reply:
The logic is that when a method accepts a context, it doesn't need the `ei` parameter, and when it accepts a buffer it needs `ei` parameter.
There are many places where we do not use the entire `ThreadContext` but we still need to read a key/value given an entry index (e.g., `binaryFind()`).
It sometimes called with `ctx.tempKey` and sometimes with `ctx.key`. So `binaryFind()` cannot accept a context as a parameter.
The other option is to make all methods accept the `ei` as a parameter, albeit it already set in the context. But this can introduce bugs that could have been easily avoided.

Reply:
Let us set some rule regarding when ThreadContext ei is used and when input parameter ei is used. writeKey for example don't have to get ei as input parameter. So the rule should be general for all usages and written somewhere in the code as a comment.

Reply:
Fixed.